File: OJ/leetcode/2462/Solution.java
import java.util.PriorityQueue;

class Solution {
    public long totalCost(int[] costs, int k, int candidates) {
        PriorityQueue<Integer> leftMinPq = new PriorityQueue<>();
        PriorityQueue<Integer> rightMinPq = new PriorityQueue<>();
        int n = costs.length;
        int l = -1;
        int r = n;
        long total = 0;
        // k sessions
        for (int i = 0; i < k; i++) {
            while ((l + 1) < r && leftMinPq.size() < candidates) {
                leftMinPq.offer(costs[++l]);
            }
            while ((r - 1) > l && rightMinPq.size() < candidates) {
                rightMinPq.offer(costs[--r]);
            }
            // hiring session
            if (leftMinPq.isEmpty()) {
                total += rightMinPq.poll();
            } else if (rightMinPq.isEmpty()) {
                total += leftMinPq.poll();
            } else if (leftMinPq.peek() <= rightMinPq.peek()) {
                // left index always less than right index
                total += leftMinPq.poll();
            } else {
                total += rightMinPq.poll();
            }
        }
        return total;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        int[] costs = {17,12,10,2,7,2,11,20,8};
        int k = 3;
        int candidates = 4;
        System.out.println("test: " + sol.totalCost(costs, k, candidates));
    }
}

File: OJ/leetcode/2462/Solution.py
from typing import List
import heapq

class Solution:
    def totalCost(self, costs: List[int], k: int, candidates: int) -> int:
        n = len(costs)
        total = 0
        if 2 * candidates + k > n:
            costs.sort()
            total = sum(costs[:k])
            return total
        leftMinHeap = []  # 建堆
        rightMinHeap = []
        for i in range(candidates):
            heapq.heappush(leftMinHeap, costs[i])
            heapq.heappush(rightMinHeap, costs[n - 1 - i])
        l = candidates
        r = n - 1 - candidates
        # k sessions
        for i in range(k):
            if leftMinHeap[0] <= rightMinHeap[0]:
                total += heapq.heappop(leftMinHeap)
                heapq.heappush(leftMinHeap, costs[l])
                l += 1
            else:
                total += heapq.heappop(rightMinHeap)
                heapq.heappush(rightMinHeap, costs[r])
                r -= 1
        return total

if __name__ == '__main__':
    sol = Solution()
    costs = [17,12,10,2,7,2,11,20,8]
    k = 3
    candidates = 4
    print("test:", sol.totalCost(costs, k, candidates))